---
title: "Visualizing Large-scale Copy Number Variation in Single-Cell RNA-Seq Expression Data"
author: 
- name: Timothy Tickle
  affiliation: &kco Klarman Cell Observatory, Broad Institute of MIT and Harvard, Cambridge, MA, USA
- name: Itay Tirosh
  affiliation: 
    - *kco
    - Weizmann Institute of Science, Rehovot, Israel
- name: Christophe Georgescu
  affiliation: *kco
- name: Maxwell Brown
  affiliation: *kco
- name: Brian Haas
  affiliation: *kco
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document: default
package: inferCNV
abstract: >
  InferCNV is used to explore tumor single cell RNA-Seq data to identify evidence for large-scale chromosomal copy number variations, such as gains or deletions of entire chromosomes or large segments of chromosomes. This is done by exploring expression intensity of genes across positions of the genome in comparison to the average or a set of reference 'normal' cells. A heatmap is generated illustrating the relative expression intensities across each chromosome, and it becomes readily apparent as to which regions of the genome are over-abundant or less-abundant as compared to normal cells (or the average, if reference normal cells are not provided).
vignette: >
  %\VignetteIndexEntry{Visualizing Large-scale Copy Number Variation in Single-Cell RNA-Seq Expression Data}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
---


# Installation
## Required dependencies

_inferCNV_ uses the _R_ packages `r CRANpkg("ape")`, `r CRANpkg("binhf")`, `r CRANpkg("fastcluster")`, `r CRANpkg("futile.logger")`, `r CRANpkg("RColorBrewer")`, `r CRANpkg("Seurat")` and imports functions from the archived `r CRANpkg("GMD")`.

<!--
[https://cran.r-project.org/web/packages/ape/index.html]
[https://cran.r-project.org/web/packages/binhf/index.html]
[https://cran.r-project.org/web/packages/fastcluster/index.html]
[https://cran.r-project.org/web/packages/futile.logger/index.html]
[https://cran.r-project.org/web/packages/RColorBrewer/index.html]
[https://cran.r-project.org/web/packages/Seurat/index.html]
[https://cran.r-project.org/src/contrib/Archive/GMD/]
-->


Use the biocLite.R script to install Bioconductor packages. To install core packages, type the following in an R command window:
```{r install, eval = FALSE}
## try http:// if https:// URLs are not supported
source("https://bioconductor.org/biocLite.R")
biocLite("inferCNV")
```

## Optional dependencies
If you want to use the interactive heatmap visualization, please also install the _R_ packages `r CRANpkg("dplyr")`, `r CRANpkg("tibble")`, `r Githubpkg("bmbroom/tsvio")` and `r Githubpkg("bmbroom/NGCHMR")`. To install optional packages, type the following in an R command window:

<!--
[https://cran.r-project.org/web/packages/dplyr/index.html]
[https://cran.r-project.org/web/packages/tibble/index.html]
[https://github.com/bmbroom/tsvio]
[https://github.com/bmbroom/NGCHMR]
-->

```{r install-optionals, eval = FALSE}
install.packages("dplyr")
install.packages("tibble")

install.packages("devtools")
devtools::install_github("bmbroom/tsvio")
devtools::install_github("bmbroom/NGCHMR", ref="stable")

```

And download the NGCHM java application by typing the following in a regular shell:
```{bash, eval = FALSE}
wget http://tcga.ngchm.net/NGCHM/ShaidyMapGen.jar
```



```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(infercnv)

```

# Running InferCNV
## Create the InferCNV Object

Reading in the raw counts matrix and meta data, populating the infercnv object

```{r}
infercnv_obj = CreateInfercnvObject(
  raw_counts_matrix="../example/oligodendroglioma_expression_downsampled.counts.matrix",
  annotations_file="../example/oligodendroglioma_annotations_downsampled.txt",
  delim="\t",
  gene_order_file="../example/gencode_downsampled.txt",
  ref_group_names=c("Microglia/Macrophage","Oligodendrocytes (non-malignant)"))

```


## Filtering genes

Removing those genes that are very lowly expressed or present in very few cells


```{r}
# filter out low expressed genes
cutoff=1
infercnv_obj <- require_above_min_mean_expr_cutoff(infercnv_obj, cutoff)

# filter out bad cells
min_cells_per_gene=3
infercnv_obj <- require_above_min_cells_ref(infercnv_obj, min_cells_per_gene=min_cells_per_gene)

## for safe keeping
infercnv_orig_filtered = infercnv_obj

#plot_mean_chr_expr_lineplot(infercnv_obj)
save('infercnv_obj', file = '../example_output/infercnv_obj.orig_filtered')

```


## Normalize each cell's counts for sequencing depth

```{r}
infercnv_obj <- infercnv:::normalize_counts_by_seq_depth(infercnv_obj)
```


## Perform Anscombe normalization

Suggested by Matan for removing noisy variation at low counts

```{r}
infercnv_obj <- infercnv:::anscombe_transform(infercnv_obj)
```
<!--save('infercnv_obj', file='../example_output/infercnv_obj.anscombe')
```-->


## Log transform the normalized counts:

```{r}
infercnv_obj <- log2xplus1(infercnv_obj)
```
<!--save('infercnv_obj', file='../example_output/infercnv_obj.log_transformed')
```-->

## Apply maximum bounds to the expression data to reduce outlier effects
```{r}
threshold = mean(abs(get_average_bounds(infercnv_obj)))
infercnv_obj <- apply_max_threshold_bounds(infercnv_obj, threshold=threshold)
```

## Initial view, before inferCNV operations:

```{r, results="hide"}
plot_cnv(infercnv_obj,
         out_dir='../example_output/', 
         output_filename='infercnv.logtransf', 
         x.range="auto", 
         title = "Before InferCNV (filtered & log2 transformed)", 
         color_safe_pal = FALSE, 
         x.center = mean(infercnv_obj@expr.data))
```


```{r, echo=FALSE}
knitr::include_graphics("../example_output/infercnv.logtransf.png")
```



## Perform smoothing across chromosomes

```{r}
infercnv_obj = smooth_by_chromosome(infercnv_obj, window_length=101, smooth_ends=TRUE)
```
<!--# save('infercnv_obj', file='../example_output/infercnv_obj.smooth_by_chr')-->

```{r}
# re-center each cell
infercnv_obj <- center_cell_expr_across_chromosome(infercnv_obj, method = "median")
```
<!--# save('infercnv_obj', file='../example_output/infercnv_obj.cells_recentered')-->


```{r, results='hide' }
plot_cnv(infercnv_obj, 
         out_dir='../example_output/',
         output_filename='infercnv.chr_smoothed', 
         x.range="auto", 
         title = "chr smoothed and cells re-centered", 
         color_safe_pal = FALSE)
```


```{r, echo=FALSE}
knitr::include_graphics("../example_output/infercnv.chr_smoothed.png")
```



## Subtract the reference values from observations, now have log(fold change) values

```{r}
infercnv_obj <- subtract_ref_expr_from_obs(infercnv_obj, inv_log=TRUE)
```
<!--
save('infercnv_obj', file='../example_output/infercnv_obj.ref_subtracted')
```-->

```{r, results="hide"}
plot_cnv(infercnv_obj,
         out_dir='../example_output/',
         output_filename='infercnv.ref_subtracted', 
         x.range="auto", 
         title="ref subtracted", 
         color_safe_pal = FALSE)
```

```{r, echo=FALSE}
knitr::include_graphics("../example_output/infercnv.ref_subtracted.png")
```


## Invert log values

Converting the log(FC) values to regular fold change values, centered at 1 (no fold change)

This is important because we want (1/2)x to be symmetrical to 1.5x, representing loss/gain of one chromosome region.

```{r}

infercnv_obj <- invert_log2(infercnv_obj)
```
<!--save('infercnv_obj', file='../example_output/infercnv_obj.inverted_log')
```-->


```{r, results="hide"}
plot_cnv(infercnv_obj, 
         out_dir='../example_output/',
         output_filename='../example_output/infercnv.inverted', 
         color_safe_pal = FALSE, 
         x.range="auto", 
         x.center=1, 
         title = "inverted log FC to FC")
```


```{r, echo=FALSE}
knitr::include_graphics("../example_output/infercnv.inverted.png")
```


## Removing noise

```{r}
infercnv_obj <- clear_noise_via_ref_mean_sd(infercnv_obj, sd_amplifier = 1.5)
```
<!--save('infercnv_obj', file='../example_output/infercnv_obj.denoised')
```-->


```{r, results="hide"}
plot_cnv(infercnv_obj,
         out_dir='../example_output/',
         output_filename='infercnv.denoised', 
         x.range="auto", 
         x.center=1, 
         title="denoised", 
         color_safe_pal = FALSE)
```

```{r, echo=FALSE}
knitr::include_graphics("../example_output/infercnv.denoised.png")
```

## Remove outlier data points

This generally improves on the visualization

```{r}
infercnv_obj = remove_outliers_norm(infercnv_obj)
```
<!--save('infercnv_obj', file="../example_output/infercnv_obj.outliers_removed")
```-->


```{r, results="hide"}

    plot_cnv(infercnv_obj, 
         out_dir='../example_output/',
         output_filename='../example_output/infercnv.outliers_removed', 
         color_safe_pal = FALSE, 
         x.range="auto", 
         x.center=1, 
         title = "outliers removed")
```


```{r, echo=FALSE}
knitr::include_graphics("../example_output/infercnv.outliers_removed.png")
```



## Find DE genes by comparing the mutant types to normal types, BASIC

Runs a t-Test comparing tumor/normal for each patient and normal sample, and masks out those genes that are not significantly DE.

```{r}

plot_data = infercnv_obj@expr.data
high_threshold = max(abs(quantile(plot_data[plot_data != 0], c(0.05, 0.95))))  

low_threshold = -1 * high_threshold 

infercnv_obj2 <- infercnv:::mask_non_DE_genes_basic(infercnv_obj, test.use = 't', center_val=1)

```


```{r, results="hide"}
plot_cnv(infercnv_obj2, 
         out_dir='../example_output/',
         output_filename='../example_output/infercnv.non-DE-genes-masked', 
         color_safe_pal = FALSE, 
         x.range=c(low_threshold, high_threshold), 
         x.center=1, 
         title = "non-DE-genes-masked")
```


```{r, echo=FALSE}
knitr::include_graphics("../example_output/infercnv.non-DE-genes-masked.png")

```

# Additional Information
## Online Documentation

For additional explanations on files, usage, and a tutorial please visit the [wiki](https://github.com/broadinstitute/inferCNV/wiki).


## TrinityCTAT
This tool is a part of the TrinityCTAT toolkit focused on leveraging the use of RNA-Seq to better understand cancer transcriptomes. To find out more please visit [TrinityCTAT](https://github.com/NCIP/Trinity_CTAT/wiki)


## Applications

This methodology was used in:

[Anoop P. Patel et al. Single-cell RNA-seq highlights intratumoral heterogeneity in primary glioblastoma. Science. 2014 Jun 20: 1396-1401](http://www.ncbi.nlm.nih.gov/pmc/articles/PMC4123637/)

[Tirosh I et al.Dissecting the multicellular ecosystem of metastatic melanoma by single-cell RNA-seq. Science. 2016 Apr 8;352(6282):189-96](http://www.ncbi.nlm.nih.gov/pubmed/27124452)




# Session info

```{r sessioninfo, echo=FALSE, tidy=TRUE, tidy.opts=list(width.cutoff=60), out.width=60}
sessionInfo()
```

